File: api/crawler/crawler.go
package crawler

import (
	"crypto/rand"
	"math/big"
	rander "math/rand"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/anaskhan96/soup"
)

// randomly picks a number from the given slice and maximum
func randomPicker(slice []string) int {
	s := rander.NewSource(time.Now().Unix())
	r := rander.New(s)
	random := r.Intn(len(slice))
	return random
}

// Getter => gets a meme from a random chosen source
func Getter() map[string]string {
	meme := make(map[string]string)

	// select a random meme website
	sites := []string{"memedroid", "gifvif"}
	randomIndex := randomPicker(sites)

	ransite := sites[randomIndex]
	if ransite == "gifvif" {
		meme["website"] = "gif-vif.com"
		meme["meme"] = GifVif()
	} else if ransite == "memedroid" {
		meme["website"] = "memedroid.com"
		meme["meme"] = MemeDroid()
	}

	// return the meme map
	return meme
}

// Compiler => compiles the return element of the two websites
func Compiler(website string) map[string]string {
	meme := make(map[string]string)

	if website == "gifvif" {
		meme["website"] = "gif-vif.com"
		meme["meme"] = GifVif()
	} else if website == "memedroid" {
		meme["website"] = "memedroid.com"
		meme["meme"] = MemeDroid()
	} else {
		meme["website"] = website
		meme["meme"] = "website not yet supported"
	}

	return meme
}

// GifVif => returns a slice of gifs from gif-vif.com
func GifVif() string {
	website := "https://www.gif-vif.com/category/funny/"

	// randomly get memes from a random page number
	randomNum := func() string {
		nbig, err := rand.Int(rand.Reader, big.NewInt(374))
		if err != nil {
			panic(err)
		}

		num := nbig.Int64()

		if num == 0 {
			num = 1
		}
		return strconv.FormatInt(num, 10)
	}

	// chooose a random gif from the links
	resp, err := soup.Get(website + randomNum())
	if err != nil {
		os.Exit(1)
	}
	doc := soup.HTMLParse(resp)
	containers := doc.FindAll("div", "id", "gif_container")
	var links []string
	for _, link := range containers {
		raw := link.Find("a")
		links = append(links, raw.Attrs()["href"])
	}

	index := randomPicker(links)

	// get the meme from the chosen link
	resp, err = soup.Get(links[index])
	if err != nil {
		os.Exit(1)
	}

	doc = soup.HTMLParse(resp)
	src := doc.Find("div", "id", "gif_div")
	gif := src.Find("source").Attrs()["src"]

	return gif
}

// MemeDroid => returns a slice of memes from memedroid.com
func MemeDroid() string {
	website := "https://www.memedroid.com/memes/random"

	resp, err := soup.Get(website)

	if err != nil {
		os.Exit(1)
	}

	doc := soup.HTMLParse(resp)

	containers := doc.FindAll("div", "class", "item-aux-container")

	var memes []string

	for _, meme := range containers {
		raw, err := meme.Find("img").Attrs()["src"]
		if err {
			if strings.HasPrefix(raw, "https://images") {
				memes = append(memes, raw)
			}
		}

	}

	index := randomPicker(memes)

	return memes[index]
}

File: api/crawler.py
import requests
from bs4 import BeautifulSoup
import random


class Crawler:
    # meme websites source
    websites = ["gifvif", "memedroid"]

    # Main default getter
    @staticmethod
    async def Getter():
        # initialize the meme dict
        meme = {}

        # select a random website
        site = random.choice(Crawler.websites)

        if site == "gifvif":
            meme["website"] = "gif-vif.com"
            meme["meme"] = Crawler.gifVif()
        elif site == "memedroid":
            meme["website"] = "memedroid.com"
            meme["meme"] = Crawler.memeDroid()
        elif site == "imgflip":
            meme["website"] = "imgflip.com"
            meme["meme"] = Crawler.imgFlip()

        # return the meme
        return meme

    # Specfic Website compiler
    @staticmethod
    async def Compiler(website):
        # initialize the meme dict
        meme = {}

        if website == "gifvif":
            meme["website"] = "gif-vif.com"
            meme["meme"] = Crawler.gifVif()
        elif website == "memedroid":
            meme["website"] = "memedroid.com"
            meme["meme"] = Crawler.memeDroid()
        elif website == "imgflip":
            meme["website"] = "imgflip.com"
            meme["meme"] = Crawler.imgFlip()

        return meme

    # GifVif website
    @staticmethod
    def gifVif():
        # initialize the website
        site = "https://www.gif-vif.com/category/funny/"
        page = random.randrange(376)  # generate a random page num

        # get the soup
        soup = BeautifulSoup(requests.get(site + str(page)).text, "html.parser")

        # get all of the links containers
        containers = soup.find("div", id="gif_container").find_all("a")

        # select one from the containers
        raw = random.choice(containers)

        # initialize meme dict
        meme = {}

        # get the gif src
        soup_rc = BeautifulSoup(requests.get(raw["href"]).text, "html.parser")

        # get the title
        meme["title"] = soup_rc.find("div", id="name_of_gif").get_text()

        # extact the src from the video tag
        meme["src"] = soup_rc.find("div", id="gif_div").find("source")["src"]

        # return the links
        return meme

    # MemeDroid website
    @staticmethod
    def memeDroid():
        # initialize the website
        site = "https://www.memedroid.com/memes/random"

        # get the soup
        soup = BeautifulSoup(requests.get(site).text, "html.parser")

        # find all containers
        containers = soup.find_all("div", class_="item-aux-container")

        memes = []

        for i in containers:
            raw = {}

            # get img src
            src = ""
            try:
                src = i.find("img")["src"]
            except Exception:
                pass

            if src.startswith("https://images"):
                # get the meme title
                raw["title"] = i.find("a", class_="item-header-title").get_text()
                raw["src"] = src

                # append the raw
                memes.append(raw)

        return random.choice(memes)

    # imgFlip (the api just returns template, so well just scrape the latest contents)
    @staticmethod
    def imgFlip():
        site = "https://imgflip.com/?sort=latest"
        
        # get the soup
        soup = BeautifulSoup(requests.get(site).text, 'html.parser')

        # get all containers
        containers = soup.find_all("div", class_="base-unit")

        # select a random from the containers
        con = random.choice(containers)

        # compile
        meme = {}
        meme["title"] = con.find("h2", class_="base-unit-title").get_text() # get the title
        meme["src"] = "https:" + con.find("img")["src"] # get the img src

        return meme